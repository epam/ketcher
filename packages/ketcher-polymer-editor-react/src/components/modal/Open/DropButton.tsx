/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { ActionButton } from 'components/shared/actionButton'
import styled from '@emotion/styled'

type Props = {
  label: string
  clickHandler: () => void
  disabled?: boolean
  caption?: string
}

const StyledActionButton = styled(ActionButton)`
  border-radius: 2px;
  text-align: center;
  width: 100%;
  font-weight: ${({ theme }) => theme.font.weight.regular};
  padding: 5px 8px;
`

const ButtonContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;

  & > span {
    padding-top: 5px;
    font-size: ${({ theme }) => theme.font.size.small};
  }
`

export const DropButton = ({ label, clickHandler, caption }: Props) => {
  return (
    <ButtonContainer>
      <StyledActionButton label={label} clickHandler={clickHandler} />
      {caption && <span>{caption}</span>}
    </ButtonContainer>
  )
}
