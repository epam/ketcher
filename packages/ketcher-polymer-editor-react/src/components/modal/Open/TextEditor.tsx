/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import styled from '@emotion/styled'

export type TextEditorProps = {
  structStr: string
  inputHandler: (str: string) => void
}

const StyledTextarea = styled.textarea`
  width: 405px;
  min-width: 32em;
  min-height: 23em;
  overflow: auto;
  white-space: pre;
  resize: both;
`

export const TextEditor = ({ structStr, inputHandler }: TextEditorProps) => {
  return (
    <>
      <StyledTextarea
        value={structStr}
        onChange={(event) => inputHandler(event.target.value)}
      />
      {/* <ClipArea */}
      {/*  focused={() => true} */}
      {/*  onCopy={() => ({ 'text/plain': structStr })} */}
      {/* /> */}
    </>
  )
}
