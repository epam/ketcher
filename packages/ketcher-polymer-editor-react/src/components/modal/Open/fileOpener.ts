/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

export type FileOpener = typeof throughFileReader;

export function fileOpener() {
  return new Promise<FileOpener>((resolve, reject) => {
    if (global.FileReader) {
      resolve(throughFileReader);
    } else {
      reject(new Error('Your browser does not support opening files locally'));
    }
  });
}

function throughFileReader(file: File) {
  return new Promise((resolve, reject) => {
    const rd = new FileReader();

    rd.onload = () => {
      const content = rd.result;
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore msClose doesn't exist in File type
      if (file.msClose) file.msClose();
      resolve(content);
    };

    rd.onerror = (event) => {
      reject(event);
    };

    rd.readAsText(file, 'UTF-8');
  });
}
