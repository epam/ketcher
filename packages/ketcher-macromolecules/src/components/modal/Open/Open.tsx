/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { Modal } from 'components/shared/modal';
import { useCallback, useEffect, useState } from 'react';
import { ViewSwitcher } from './ViewSwitcher';
import { ActionButton } from 'components/shared/actionButton';
import { FileOpener, fileOpener } from './fileOpener';
import {
  ChemicalMimeType,
  KetSerializer,
  StructService,
  CoreEditor,
  KetcherLogger,
  EditorHistory,
  SequenceMode,
} from 'ketcher-core';
import { IndigoProvider } from 'ketcher-react';
import { RequiredModalProps } from '../modalContainer';
import { OpenFileWrapper } from './Open.styles';
import { Loader, StyledDropdown, stylesForExpanded } from '../save/Save.styles';
import { LoadingCircles } from './AnalyzingFile/LoadingCircles';
import { useAppDispatch } from 'hooks';
import { openErrorModal } from 'state/modal';
import { AnyAction, Dispatch } from 'redux';
import styled from '@emotion/styled';
import { Option } from 'components/shared/dropDown/dropDown';

export interface Props {
  onClose: () => void;
  isModalOpen: boolean;
}

const OpenModal = styled(Modal)(
  ({ modalWidth }) => `
    .MuiPaper-root {
      width: ${modalWidth};
    }`,
);

const OpenFooter = styled.div({
  width: '100%',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-between',
});

const FooterSelectorContainer = styled.div({
  height: '24px',
  fontSize: '12px',
});

const FooterFormatSelector = styled(StyledDropdown)({
  width: '180px',
});

const FooterSequenceSelector = styled(StyledDropdown)({
  width: '140px',
  margin: '0 8px',
});

const FooterButton = styled(ActionButton)({
  margin: '0 8px',
  width: '100px',
});

const KET = 'ket';
const SEQ = 'seq';
const RNA = 'rna';

const options: Array<Option> = [
  { id: 'ket', label: 'Ket' },
  { id: 'mol', label: 'MDL Molfile V3000' },
  { id: 'seq', label: 'Sequence' },
];

const sequenceOptions: Array<Option> = [
  { id: 'rna', label: 'RNA' },
  { id: 'dna', label: 'DNA' },
  { id: 'peptide', label: 'Peptide' },
];

const inputFormats = {
  ket: 'chemical/x-indigo-ket',
  mol: 'chemical/x-mdl-molfile',
  rna: 'chemical/x-rna-sequence',
  dna: 'chemical/x-dna-sequence',
  peptide: 'chemical/x-peptide-sequence',
};

export const MODAL_STATES = {
  openOptions: 'openOptions',
  textEditor: 'textEditor',
} as const;

export type MODAL_STATES_VALUES =
  typeof MODAL_STATES[keyof typeof MODAL_STATES];

const addToCanvas = ({
  ketSerializer,
  editor,
  struct,
}: {
  ketSerializer: KetSerializer;
  editor: CoreEditor;
  struct: string;
}) => {
  const deserialisedKet = ketSerializer.deserializeToDrawingEntities(struct);

  if (!deserialisedKet) {
    throw new Error('Error during parsing file');
  }

  deserialisedKet.drawingEntitiesManager.centerMacroStructure();
  const modelChanges = deserialisedKet.drawingEntitiesManager.mergeInto(
    editor.drawingEntitiesManager,
  );
  const editorHistory = new EditorHistory(editor);
  const isSequenceMode = editor.mode instanceof SequenceMode;

  editor.renderersContainer.update(modelChanges);
  editorHistory.update(modelChanges);

  if (isSequenceMode) {
    modelChanges.setUndoOperationReverse();
    editor.events.selectMode.dispatch({
      mode: 'flex-layout-mode',
      mergeWithLatestHistoryCommand: true,
    });
    editor.events.selectMode.dispatch({
      mode: 'sequence-layout-mode',
      mergeWithLatestHistoryCommand: true,
    });
  }
};

// TODO: replace after the implementation of the function for processing the structure from the file
const onOk = async ({
  struct,
  formatSelection,
  sequenceSelection,
  onCloseCallback,
  setIsLoading,
  dispatch,
}: {
  struct: string;
  formatSelection: string;
  sequenceSelection: string;
  onCloseCallback: () => void;
  setIsLoading: (isLoading: boolean) => void;
  dispatch: Dispatch<AnyAction>;
}) => {
  const isKet = formatSelection === KET;
  const isSeq = formatSelection === SEQ;
  const ketSerializer = new KetSerializer();
  const editor = CoreEditor.provideEditorInstance();
  let inputFormat;
  let fileData = struct;

  const showParsingError = (stringError) => {
    const errorMessage = 'Convert error! ' + stringError;
    dispatch(
      openErrorModal({
        errorMessage,
        errorTitle: isSeq ? 'Unsupported symbols' : '',
      }),
    );
  };

  if (isKet) {
    try {
      addToCanvas({ struct, ketSerializer, editor });
      onCloseCallback();
    } catch (e) {
      showParsingError('Error during file parsing.');
    }
    return;
  } else if (isSeq) {
    inputFormat = inputFormats[sequenceSelection];
    fileData = fileData.toUpperCase();
  } else {
    inputFormat = inputFormats[formatSelection];
  }
  const indigo = IndigoProvider.getIndigo() as StructService;

  try {
    setIsLoading(true);

    const ketStruct = await indigo.convert({
      struct: fileData,
      output_format: ChemicalMimeType.KET,
      input_format: inputFormat,
    });
    addToCanvas({ struct: ketStruct.struct, ketSerializer, editor });
    onCloseCallback();
  } catch (error) {
    const stringError =
      typeof error === 'string' ? error : JSON.stringify(error);
    showParsingError(stringError);
    KetcherLogger.error(error);
  } finally {
    setIsLoading(false);
  }
};
const isAnalyzingFile = false;
const errorHandler = (error) => console.log(error);

const Open = ({ isModalOpen, onClose }: RequiredModalProps) => {
  const dispatch = useAppDispatch();
  const [structStr, setStructStr] = useState<string>('');
  const [fileName, setFileName] = useState<string>('');
  const [isLoading, setIsLoading] = useState(false);
  const [opener, setOpener] = useState<
    { chosenOpener: FileOpener } | undefined
  >();
  const [currentState, setCurrentState] = useState<MODAL_STATES_VALUES>(
    MODAL_STATES.openOptions,
  );
  const [formatSelection, setFormatSelection] = useState(KET);
  const [sequenceSelection, setSequenceSelection] = useState(RNA);

  useEffect(() => {
    const splittedFilenameByDot = fileName?.split('.');
    const fileExtension =
      splittedFilenameByDot[splittedFilenameByDot.length - 1];

    if (fileExtension) {
      const option = options.find((el) => el.id === fileExtension);
      const id = option?.id ? option.id : SEQ;
      setFormatSelection(id);
    }
  }, [fileName]);

  useEffect(() => {
    fileOpener().then((chosenOpener) => {
      setOpener({ chosenOpener });
    });
  }, []);

  const onCloseCallback = useCallback(() => {
    setCurrentState(MODAL_STATES.openOptions);
    setStructStr('');
    setFormatSelection(KET);
    setSequenceSelection(RNA);
    onClose();
  }, [onClose]);

  const onFileLoad = (files: File[]) => {
    const onLoad = (fileContent) => {
      setStructStr(fileContent);
      setCurrentState(MODAL_STATES.textEditor);
    };
    const onError = () => errorHandler('Error processing file');

    setFileName(files[0].name);
    opener?.chosenOpener(files[0]).then(onLoad, onError);
  };

  const copyHandler = () => {
    onOk({
      struct: structStr,
      formatSelection,
      sequenceSelection,
      onCloseCallback,
      setIsLoading,
      dispatch,
    });
  };

  const openHandler = () => {
    const editor = CoreEditor.provideEditorInstance();
    const history = new EditorHistory(editor);
    const modelChanges = editor.drawingEntitiesManager.deleteAllEntities();

    history.update(modelChanges);
    editor.renderersContainer.update(modelChanges);

    onOk({
      struct: structStr,
      formatSelection,
      sequenceSelection,
      onCloseCallback,
      setIsLoading,
      dispatch,
    });
  };

  const renderFooter = () => (
    <OpenFooter>
      <FooterSelectorContainer>
        <FooterFormatSelector
          options={options}
          currentSelection={formatSelection}
          selectionHandler={setFormatSelection}
          customStylesForExpanded={stylesForExpanded}
          key={formatSelection}
        />
        {formatSelection === SEQ ? (
          <FooterSequenceSelector
            options={sequenceOptions}
            currentSelection={sequenceSelection}
            selectionHandler={setSequenceSelection}
            customStylesForExpanded={stylesForExpanded}
            key={sequenceSelection}
          />
        ) : null}
      </FooterSelectorContainer>
      <div>
        <FooterButton
          key="openButton"
          disabled={!structStr}
          clickHandler={openHandler}
          label="Open as New"
          styleType="secondary"
        />
        <FooterButton
          key="copyButton"
          disabled={!structStr}
          clickHandler={copyHandler}
          label="Add to Canvas"
          title="Structure will be loaded as fragment and added to Clipboard"
          data-testid="add-to-canvas-button"
        />
      </div>
    </OpenFooter>
  );

  return (
    <OpenModal
      isOpen={isModalOpen}
      title="Open Structure"
      onClose={onCloseCallback}
      modalWidth={currentState === MODAL_STATES.textEditor ? '600px' : ''}
    >
      <Modal.Content>
        <OpenFileWrapper currentState={currentState}>
          <ViewSwitcher
            isAnalyzingFile={isAnalyzingFile}
            fileName={fileName}
            currentState={currentState}
            states={MODAL_STATES}
            selectClipboard={() => setCurrentState(MODAL_STATES.textEditor)}
            fileLoadHandler={onFileLoad}
            errorHandler={errorHandler}
            value={structStr}
            inputHandler={setStructStr}
          />
          {isLoading && (
            <Loader>
              <LoadingCircles />
            </Loader>
          )}
        </OpenFileWrapper>
      </Modal.Content>
      {currentState === MODAL_STATES.textEditor && !isAnalyzingFile ? (
        <Modal.Footer withborder="true">{renderFooter()}</Modal.Footer>
      ) : (
        <></>
      )}
    </OpenModal>
  );
};
export { Open };
