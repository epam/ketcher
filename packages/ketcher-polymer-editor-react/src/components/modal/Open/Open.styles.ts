import styled from '@emotion/styled';
import { MODAL_STATES, MODAL_STATES_VALUES } from './Open';
import { ActionButton } from 'components/shared/actionButton';

interface OpenFileWrapperProps {
  currentState: MODAL_STATES_VALUES;
}

export const OpenFileWrapper = styled.div<OpenFileWrapperProps>`
  position: relative;
  padding: ${({ currentState }) =>
    currentState === MODAL_STATES.openOptions ? '10px 12px' : '0'};
`;

export const CancelButton = styled(ActionButton)`
  margin-right: auto;
`;
