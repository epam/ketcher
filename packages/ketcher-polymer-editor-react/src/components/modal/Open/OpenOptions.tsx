/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { FC } from 'react'
import { Icon } from 'components/shared/icon'
import styled from '@emotion/styled'
import { FileDrop } from 'components/modal/Open/FileDrop'
import { DropButton } from 'components/modal/Open/DropButton'

const ICON_NAMES = {
  PASTE: 'capital-t',
  FILE: 'arrow-up'
}

export type OpenOptionsProps = {
  selectClipboard: () => void
  errorHandler: (err: string) => void
  fileLoadHandler: (files: File[]) => void
  isRecognizeDisabled: boolean
}

const RootContainer = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: flex-start;
  margin-bottom: 20px;
  & > * {
    margin-right: 18px;
  }

  & :last-child {
    margin-right: 0;
  }
`

const DropContainer = styled.div`
  width: 152px;
  height: 112px;
  border: 1px solid #aeaeae;
  border-radius: 4px;
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;
  box-sizing: border-box;
  position: relative;
  padding: 15px 6px;

  & > p {
    margin-top: 6px;
    text-align: center;
  }

  svg {
    fill: ${({ theme }) => theme.color.button.primary.active};
  }
`

export const OpenOptions: FC<OpenOptionsProps> = ({
  selectClipboard,
  fileLoadHandler,
  errorHandler
}) => {
  return (
    <RootContainer>
      <DropContainer>
        <DropButton
          clickHandler={selectClipboard}
          label="Paste from Clipboard"
        />
        <Icon name={ICON_NAMES.PASTE} />
      </DropContainer>
      <DropContainer>
        <FileDrop
          onDropAccepted={fileLoadHandler}
          onDropRejected={(e) => errorHandler(`Unable to accept file(s). ${e}`)}
          buttonLabel="Open from file"
          textLabel="or drag file here"
          iconName={ICON_NAMES.FILE}
        />
      </DropContainer>
    </RootContainer>
  )
}
