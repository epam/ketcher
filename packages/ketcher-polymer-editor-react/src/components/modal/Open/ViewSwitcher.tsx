/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { AnalyzingFile } from './AnalyzingFile'

import { OpenOptions } from './OpenOptions'
import { TextEditor } from './TextEditor'

// type ViewStates = {
//   idle: string
//   textEditor: string
// }

// type SwitchProps = {
//   currentState: string
//   states: ViewStates
//   isAnalyzingFile: boolean
// } & OpenOptionsProps &
//   TextEditorProps &
//   AnalyzingFileProps

export const ViewSwitcher = (props) => {
  if (props.isAnalyzingFile) {
    return <AnalyzingFile fileName={props.fileName} />
  } else {
    switch (props.currentState) {
      case props.states.idle:
        return <OpenOptions {...props} />
      case props.states.textEditor:
        return <TextEditor {...props} />
      default:
        return null
    }
  }
}
