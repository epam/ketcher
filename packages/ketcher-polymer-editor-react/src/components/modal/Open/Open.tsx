/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/
import { Modal } from 'components/shared/modal';
import { useCallback, useEffect, useState } from 'react';
import { ViewSwitcher } from './ViewSwitcher';
import { ActionButton } from 'components/shared/actionButton';
import { FileOpener, fileOpener } from './fileOpener';

export interface Props {
  onClose: () => void;
  isModalOpen: boolean;
}

const MODAL_STATES = {
  openOptions: 'openOptions',
  textEditor: 'textEditor',
};

// TODO: replace after the implementation of the function for processing the structure from the file
const onOk = ({ struct, fragment }) => {
  if (fragment) {
    console.log('add fragment');
  }
  console.log(struct);
};
const isAnalyzingFile = false;
const errorHandler = (error) => console.log(error);

const Open = ({ isModalOpen, onClose }: Props) => {
  const [structStr, setStructStr] = useState<string>('');
  const [fileName, setFileName] = useState<string>('');
  const [opener, setOpener] = useState<
    { chosenOpener: FileOpener } | undefined
  >();
  const [currentState, setCurrentState] = useState(MODAL_STATES.openOptions);

  useEffect(() => {
    fileOpener().then((chosenOpener) => {
      setOpener({ chosenOpener });
    });
  }, []);

  const onCloseCallback = useCallback(() => {
    setCurrentState(MODAL_STATES.openOptions);
    setStructStr('');
    onClose();
  }, [onClose]);

  const onFileLoad = (files: File[]) => {
    const onLoad = (fileContent) => {
      setStructStr(fileContent);
      setCurrentState(MODAL_STATES.textEditor);
    };
    const onError = () => errorHandler('Error processing file');

    setFileName(files[0].name);
    opener?.chosenOpener(files[0]).then(onLoad, onError);
  };

  const copyHandler = () => {
    onOk({ struct: structStr, fragment: true });
    onCloseCallback();
  };

  const openHandler = () => {
    onOk({ struct: structStr, fragment: false });
    onCloseCallback();
  };

  const getButtons = () => {
    if (currentState === MODAL_STATES.textEditor && !isAnalyzingFile) {
      return [
        <ActionButton
          key="copyButton"
          disabled={!structStr}
          clickHandler={copyHandler}
          label="Add to Canvas"
          title="Structure will be loaded as fragment and added to Clipboard"
          styleType="secondary"
        />,
        <ActionButton
          key="openButton"
          disabled={!structStr}
          clickHandler={openHandler}
          label="Open as New Project"
        />,
      ];
    } else {
      return [];
    }
  };

  return (
    <Modal
      isOpen={isModalOpen}
      title="Open Structure"
      onClose={onCloseCallback}
    >
      <Modal.Content>
        <ViewSwitcher
          isAnalyzingFile={isAnalyzingFile}
          fileName={fileName}
          currentState={currentState}
          states={MODAL_STATES}
          selectClipboard={() => setCurrentState(MODAL_STATES.textEditor)}
          fileLoadHandler={onFileLoad}
          errorHandler={errorHandler}
          struct={structStr}
          inputHandler={setStructStr}
        />
      </Modal.Content>
      <Modal.Footer>{getButtons()}</Modal.Footer>
    </Modal>
  );
};
export { Open };
