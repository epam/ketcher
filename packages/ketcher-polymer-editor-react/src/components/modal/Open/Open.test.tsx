/****************************************************************************
 * Copyright 2021 EPAM Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ***************************************************************************/

import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { Open } from './Open';

const mockProps = {
  isModalOpen: true,
  onClose: jest.fn(),
};

// eslint-disable-next-line @typescript-eslint/no-empty-function
jest.spyOn(React, 'useEffect').mockImplementation(() => {});

describe('Open component', () => {
  it('should render correctly', () => {
    expect(
      render(withThemeProvider(<Open {...mockProps} />)),
    ).toMatchSnapshot();
  });
  it('paste from clipboard', () => {
    const mockTypedText = 'more typed text';
    render(withThemeProvider(<Open {...mockProps} />));
    const clipboardButton = screen.getByRole('button', {
      name: 'Paste from Clipboard',
    });
    fireEvent.click(clipboardButton);

    const clipboardTextarea = screen.getByRole('textbox');
    fireEvent.change(clipboardTextarea, { target: { value: mockTypedText } });
    expect(clipboardTextarea).toBeInTheDocument();
    expect(clipboardTextarea).toHaveValue(mockTypedText);

    const newProjectButton = screen.getByRole('button', {
      name: 'Open as New Project',
    });
    fireEvent.click(newProjectButton);
    expect(mockProps.onClose).toHaveBeenCalled();
  });
});
