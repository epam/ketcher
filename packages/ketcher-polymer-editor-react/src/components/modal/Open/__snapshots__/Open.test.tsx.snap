// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Open component should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="padding-right: 1024px; overflow: hidden;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="MuiModal-root MuiDialog-root css-zw3mfo-MuiModal-root-MuiDialog-root"
      role="presentation"
    >
      <div
        aria-hidden="true"
        class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
        style="opacity: 0.4; background: rgb(0, 86, 98); webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      />
      <div
        data-test="sentinelStart"
        tabindex="0"
      />
      <div
        class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
        role="presentation"
        style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        tabindex="-1"
      >
        <div
          aria-labelledby=":r0:"
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd css-12rl710-MuiPaper-root-MuiDialog-paper"
          role="dialog"
          style="min-width: 20vw; background: rgb(242, 242, 242); border-radius: 8px; color: rgb(51, 51, 51);"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-1x5ms5d-MuiTypography-root-MuiDialogTitle-root"
            id=":r0:"
          >
            <div
              class="css-lvyu5j"
            >
              Open Structure
            </div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              title="Close window"
              type="button"
            >
              <icon-close.svg
                fill="#525252"
                role="img"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </h2>
          <div
            class="MuiDialogContent-root css-1n1tez8-MuiDialogContent-root"
          >
            <div
              class="css-9brf46"
            >
              <div
                class="css-2jv7sj"
              >
                <button
                  class="MuiButtonBase-root css-1u2dyik-MuiButtonBase-root"
                  role="button"
                  tabindex="0"
                  title="Paste from Clipboard"
                  type="button"
                >
                  Paste from Clipboard
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
                <icon-capital-t.svg
                  fill="#525252"
                  role="img"
                />
              </div>
              <div
                class="css-2jv7sj"
              >
                <div
                  role="button"
                  style="width: 100%; height: 100%; display: flex; align-items: center; flex-direction: column; justify-content: space-around;"
                  tabindex="0"
                >
                  <input
                    autocomplete="off"
                    style="display: none;"
                    tabindex="-1"
                    type="file"
                  />
                  <div
                    class="css-1gnig72"
                  >
                    <button
                      class="MuiButtonBase-root css-1u2dyik-MuiButtonBase-root"
                      role="button"
                      tabindex="0"
                      title="Open from file"
                      type="button"
                    >
                      Open from file
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                    <span>
                      or drag file here
                    </span>
                  </div>
                  <icon-arrow-upward.svg
                    fill="#525252"
                    role="img"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="MuiDialogActions-root MuiDialogActions-spacing css-1codpvf-MuiDialogActions-root"
          />
        </div>
      </div>
      <div
        data-test="sentinelEnd"
        tabindex="0"
      />
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
